File: iris_analysis.py
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import seaborn as sns
import pandas as pd
from sklearn.datasets import load_iris

sns.set_theme(style="whitegrid", palette="muted", font_scale=1.2)
COLORS = {"setosa": "#4C72B0", "versicolor": "#55A868", "virginica": "#C44E52"}

iris = load_iris(as_frame=True)
df = iris.frame
df.columns = [c.replace(" (cm)", "").replace(" ", "_") for c in df.columns]
df["species"] = df["target"].map(dict(enumerate(iris.target_names)))
df = df.drop(columns="target")

print("\n=== Iris Dataset — Summary Statistics ===\n")
print(df.groupby("species").describe().T.to_string())
print()

# 1. Pair plot
pair = sns.pairplot(
    df,
    hue="species",
    palette=COLORS,
    plot_kws={"alpha": 0.7, "s": 60, "edgecolor": "white", "linewidth": 0.5},
    diag_kind="kde",
    corner=False,
)
pair.figure.suptitle("Iris Dataset — Scatter Plot Matrix", y=1.02, fontsize=16, fontweight="bold")
pair.savefig("pair_plot.png", dpi=150, bbox_inches="tight")
plt.close()
print("Saved pair_plot.png")

# 2. Box plots
features = ["sepal_length", "sepal_width", "petal_length", "petal_width"]
fig, axes = plt.subplots(2, 2, figsize=(12, 9))
fig.suptitle("Iris Dataset — Measurements by Species", fontsize=16, fontweight="bold", y=1.01)
for ax, feat in zip(axes.flat, features):
    sns.boxplot(
        data=df, x="species", y=feat, hue="species", palette=COLORS, legend=False,
        width=0.5, linewidth=1.2, flierprops={"marker": "o", "markersize": 5, "alpha": 0.6},
        ax=ax,
    )
    sns.stripplot(data=df, x="species", y=feat, hue="species", palette=COLORS, legend=False, size=3, alpha=0.4, jitter=True, ax=ax)
    ax.set_title(feat.replace("_", " ").title(), fontsize=13)
    ax.set_xlabel("")
plt.tight_layout()
fig.savefig("box_plots.png", dpi=150, bbox_inches="tight")
plt.close()
print("Saved box_plots.png")

# 3. Correlation heatmap
fig, ax = plt.subplots(figsize=(7, 6))
corr = df[features].corr()
mask = pd.DataFrame(False, index=corr.index, columns=corr.columns)
for i in range(len(mask)):
    for j in range(i):
        mask.iloc[i, j] = True
sns.heatmap(
    corr, mask=mask, annot=True, fmt=".2f", cmap="coolwarm",
    vmin=-1, vmax=1, linewidths=0.5, square=True,
    cbar_kws={"shrink": 0.8}, ax=ax,
)
ax.set_title("Iris Dataset — Feature Correlation Matrix", fontsize=15, fontweight="bold", pad=14)
fig.savefig("correlation_heatmap.png", dpi=150, bbox_inches="tight")
plt.close()
print("Saved correlation_heatmap.png")

# 4. Violin plot — petal length
fig, ax = plt.subplots(figsize=(9, 6))
sns.violinplot(
    data=df, x="species", y="petal_length", hue="species", palette=COLORS, legend=False,
    inner="quartile", linewidth=1.2, ax=ax,
)
sns.stripplot(data=df, x="species", y="petal_length", hue="species", palette=COLORS, legend=False, size=4, alpha=0.5, jitter=True, ax=ax)
ax.set_title("Iris Dataset — Petal Length by Species", fontsize=15, fontweight="bold")
ax.set_xlabel("Species", fontsize=12)
ax.set_ylabel("Petal Length (cm)", fontsize=12)
fig.savefig("violin_petal_length.png", dpi=150, bbox_inches="tight")
plt.close()
print("Saved violin_petal_length.png")

# 5. Grouped bar plot — mean ± SEM by species
import numpy as np

def sem(x):
    return x.std(ddof=1) / np.sqrt(len(x))

grp = df.groupby("species")[features]
means_df = grp.mean().add_suffix("_mean")
sems_df  = grp.agg(sem).add_suffix("_sem")
summary  = means_df.join(sems_df)

species_list = list(COLORS.keys())
x = np.arange(len(features))
width = 0.22
offsets = [-width, 0, width]

fig, ax = plt.subplots(figsize=(11, 6))
for i, sp in enumerate(species_list):
    means = [summary.loc[sp, f"{f}_mean"] for f in features]
    sems  = [summary.loc[sp, f"{f}_sem"]  for f in features]
    bars = ax.bar(x + offsets[i], means, width,
                  label=sp.capitalize(), color=COLORS[sp],
                  alpha=0.85, edgecolor="white", linewidth=0.6)
    ax.errorbar(x + offsets[i], means, yerr=sems,
                fmt="none", color="#333333", capsize=4, capthick=1.2, linewidth=1.2)

ax.set_xticks(x)
ax.set_xticklabels([f.replace("_", " ").title() for f in features], fontsize=12)
ax.set_ylabel("Mean Value (cm)", fontsize=12)
ax.set_title("Iris Dataset — Mean Measurements by Species (± SEM)",
             fontsize=14, fontweight="bold", pad=12)
ax.legend(title="Species", fontsize=11, title_fontsize=11)
ax.set_ylim(0, ax.get_ylim()[1] * 1.12)
sns.despine()
plt.tight_layout()
fig.savefig("figure6_errorbars.png", dpi=150, bbox_inches="tight")
plt.close()
print("Saved figure6_errorbars.png")

print("\nAll figures saved.")

File: iris_analysis.R
library(ggplot2)
library(GGally)
library(reshape2)

theme_set(theme_minimal(base_size = 13))

COLORS <- c("setosa" = "#4C72B0", "versicolor" = "#55A868", "virginica" = "#C44E52")

# Summary statistics
cat("\n=== Iris Dataset — Summary Statistics ===\n\n")
print(summary(iris))
cat("\nBy species:\n")
for (sp in levels(iris$Species)) {
  cat("\n--", sp, "--\n")
  print(summary(iris[iris$Species == sp, 1:4]))
}

# 1. Scatter plot matrix
p1 <- ggpairs(
  iris,
  columns = 1:4,
  aes(color = Species, alpha = 0.7),
  upper = list(continuous = wrap("cor", size = 3.5)),
  lower = list(continuous = wrap("points", size = 1.5, alpha = 0.6)),
  diag  = list(continuous = wrap("densityDiag", alpha = 0.5)),
  title = "Iris Dataset — Scatter Plot Matrix"
) +
  scale_color_manual(values = COLORS) +
  scale_fill_manual(values = COLORS)

ggsave("pair_plot.png", plot = p1, width = 10, height = 9, dpi = 150)
cat("Saved pair_plot.png\n")

# 2. Box plots — faceted by measurement
iris_long <- reshape2::melt(iris, id.vars = "Species",
                            variable.name = "Measurement", value.name = "Value")

p2 <- ggplot(iris_long, aes(x = Species, y = Value, fill = Species)) +
  geom_boxplot(width = 0.5, outlier.shape = 21, outlier.size = 2, alpha = 0.8) +
  geom_jitter(aes(color = Species), width = 0.15, size = 1.2, alpha = 0.4) +
  facet_wrap(~Measurement, scales = "free_y", ncol = 2) +
  scale_fill_manual(values = COLORS) +
  scale_color_manual(values = COLORS) +
  labs(title = "Iris Dataset — Measurements by Species", x = NULL, y = "Value (cm)") +
  theme(legend.position = "none",
        strip.text = element_text(face = "bold"),
        plot.title = element_text(face = "bold", size = 15))

ggsave("box_plots.png", plot = p2, width = 10, height = 8, dpi = 150)
cat("Saved box_plots.png\n")

# 3. Correlation heatmap
corr_mat <- round(cor(iris[, 1:4]), 2)
corr_df  <- reshape2::melt(corr_mat)

p3 <- ggplot(corr_df, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile(color = "white", linewidth = 0.5) +
  geom_text(aes(label = value), size = 4.5, fontface = "bold") +
  scale_fill_gradient2(low = "#C44E52", mid = "white", high = "#4C72B0",
                       midpoint = 0, limits = c(-1, 1), name = "r") +
  labs(title = "Iris Dataset — Feature Correlation Matrix", x = NULL, y = NULL) +
  coord_fixed() +
  theme(axis.text.x = element_text(angle = 30, hjust = 1),
        plot.title  = element_text(face = "bold", size = 15),
        panel.grid  = element_blank())

ggsave("correlation_heatmap.png", plot = p3, width = 7, height = 6, dpi = 150)
cat("Saved correlation_heatmap.png\n")

# 4. Violin plot — Petal.Length by Species
p4 <- ggplot(iris, aes(x = Species, y = Petal.Length, fill = Species, color = Species)) +
  geom_violin(alpha = 0.6, trim = FALSE, linewidth = 0.8) +
  geom_boxplot(width = 0.1, fill = "white", alpha = 0.8, outlier.shape = NA, linewidth = 0.7) +
  geom_jitter(width = 0.08, size = 1.8, alpha = 0.5) +
  scale_fill_manual(values = COLORS) +
  scale_color_manual(values = COLORS) +
  labs(title = "Iris Dataset — Petal Length by Species",
       x = "Species", y = "Petal Length (cm)") +
  theme(legend.position = "none",
        plot.title = element_text(face = "bold", size = 15))

ggsave("violin_petal_length.png", plot = p4, width = 8, height = 6, dpi = 150)
cat("Saved violin_petal_length.png\n")

# 5. Grouped bar plot — mean ± SEM by species
iris_long2 <- reshape2::melt(iris, id.vars = "Species",
                             variable.name = "Measurement", value.name = "Value")

p5 <- ggplot(iris_long2, aes(x = Measurement, y = Value, fill = Species)) +
  stat_summary(fun = mean, geom = "bar",
               position = position_dodge(width = 0.7), width = 0.65,
               alpha = 0.85, color = "white", linewidth = 0.4) +
  stat_summary(fun.data = mean_se, geom = "errorbar",
               position = position_dodge(width = 0.7), width = 0.25,
               color = "#333333", linewidth = 0.9) +
  scale_fill_manual(values = COLORS) +
  scale_x_discrete(labels = c("Sepal.Length" = "Sepal Length",
                               "Sepal.Width"  = "Sepal Width",
                               "Petal.Length" = "Petal Length",
                               "Petal.Width"  = "Petal Width")) +
  labs(title = "Iris Dataset — Mean Measurements by Species (± SEM)",
       x = NULL, y = "Mean Value (cm)", fill = "Species") +
  theme(plot.title  = element_text(face = "bold", size = 14),
        axis.text.x = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.text  = element_text(size = 10))

ggsave("figure5_errorbars.png", plot = p5, width = 10, height = 6, dpi = 150)
cat("Saved figure5_errorbars.png\n")

cat("\nAll figures saved.\n")
